### PR TITLE
Update VotingReputation for Multicall

### LINKED ISSUE #1128: Address issues with making batch changes to expenditures
https://www.figma.com/file/A6p0TCq8T1I8kYiT4q5UeM/Handling-changes-on-expenditures?node-id=0-1&t=NTpfPsx3wNEgQyEg-0

With the addition of multicall, some of the assumption made in the `VotingReputation` contract no longer hold. We propose the following changes:

- [ ] Restrict multicall functionality to one level (i.e. no more nested multicalls)
- [ ] Restrict expenditure locking behavior to expenditure-wide, no more slot-level locking
- [ ] Restrict motions on expenditures to one-at-a-time, if a motion is staked while an existing one is ongoing, it is a no-op
- [ ] Upon the resolution of a motion to edit an expenditure, the global claim delay is reset to the default value

### PR DESCRIPTION
**Pull request by kronosapiens**
Closes #1128 

Add proper handling for multicall-based motions. Specifically, adds handling for parsing the constituent multicall actions and to perform proper branching based on the contents.

Constituent actions fall into three categories:
- root-level actions,
- domain-level actions, and
- expenditure-level actions.
 
Actions can be combined arbitrarily, subject to the following requirements:
- all actions act in the same domain, and
- all expenditure actions act on the same expenditure.

Backwards-compatibility is implemented as follows:
- If the extension is upgrade while there is an ongoing motion for the expenditure, it can be completed
- New motions cannot be made on the expenditure while "old" motions are still active
- Once all "old" motions are finalized, new motions may be made for an expenditure

This PR makes the following additional changes:
- breaks the voting contract into two contracts due to finally hitting the contract size limit
- sets an expenditure's `finalizedTimestamp` to `block.timestamp` on finalization, in effect resetting the claim delay
- removes slot-level expenditure locking. Locking now operates only on whole expenditures
- relatedly, replaces the semaphore-based expenditure locking with a mutex, i.e. only one motion can be active at a given time for an expenditure. Any additional motions will be no-ops
- fixes a bug where voting power was relevant even in the root domain, whereas the spec says root motions should not consider voting history

### CODE REVIEW

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:278]: It does. Putting `require` in a view function makes testing more challenging, so I opted to move the requires into the calling function, and simply return UINT256_MAX here.

File: contracts/extensions/votingReputation/VotingReputation.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputation.sol:122]: Haven't decided 100%. We need to reference the `expenditureId`, but we can always look it up from the action. Storing `multicallSig` would let us pull it out as needed, which is only twice.

Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputation.sol]: You're right, in the end we can get away with storing just the sig.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol]: I'm never a fan of changing the code for the benefits of the test, but regardless, I'd argue that this comment is therefore inaccurate, and would expect it to say something like `... set to UINT256_MAX which represents an invalid multicall and should be handled as such wherever this function is called`

Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol]: I have added a test that I believe should pass, and it seems like the simple case that has slipped through has only done so because of this pattern.

Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol]: You're right, I missed that.

**Review by area — COMMENTED** [review]
Been finding this this very hard work to review and try and tease out all the potential regimes that might be hit. Here are my comments for now.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:246]: Are you sure about the second case? The end of the `else` clause is:

```
        if (!isExpenditureSig(summary.sig)) {
          summary.sig = sig;
        }
```

Which would return the last function signature if no expenditure signatures were seen.

The comment says as much:

```
        // If no expenditure signatures have been seen, we record the latest signature
```

Also, if the action is not a multicall, then we parse the single action, returning the same summary as if it were a multicall of one action.

File: contracts/extensions/votingReputation/VotingReputation.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputation.sol:269]: Hmm. That's valid I suppose, although I expect this would be an extremely unlikely case. Let me think about it for a bit.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:246]: Your comment would be more accurately phrased as:

- If passed an action that is a multicall, returns the first expenditure sig inside a multicall
- If passed an action that is a multicall, and there are no expenditure sigs inside, returns  the final function sig
- If passed an action that is a not multicall, returns the summary of the single action

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:1]: Oh, no, maybe they do in terms of the inheritance graph. That's not ideal. Maybe we need to change how we structure extensions to accommodate this in a gas-efficient way. Maybe ExtensionStorage, inherited by everything, and ColonyExtension which inherits ExtensionStorage?

File: contracts/extensions/votingReputation/VotingReputationStaking.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStaking.sol:226]: I did initially, but I thought it would be a suboptimal implementation, since some of that data is already stored on the motion struct (i.e. the `domainId`). I figured it would be easier to look up the `expenditureId` the few times it's needed, and so wouldn't be worth storing.

File: contracts/extensions/votingReputation/VotingReputation.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputation.sol:269]: Hmm. In our case, the only times this branching is relevant is if we're dealing with an expenditure-related action. So we have a few scenarios:

- An old motion is created with a function chosen to have a 0x0 signature. In that case, we wouldn't enter this branch because it's not an expenditure-related function.
- A new motion is created with a function chosen to have a 0x0 signature. Same logic applies.
- A multicall is created with an expenditure function and the final function chosen to have a 0x0 signature. In that case, the sig will be the expenditure signature.

I don't see a scenario where an expenditure-related motion could end up with a sig value of 0x0 which isn't an upgrade, and if someone attempted to engineer a motion to have such a signature, these branches would not be hit. What do you think? It's possible I've made a mistake in my analysis.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:246]: Ok, I see what you're saying. I think I was able to get away with it during my implementation because it's called in contexts where we either already know we're dealing with an expenditure (for example by previously checking `motion.sig`) or the result of `isExpenditureSig(getSig(action))` will be false regardless of whether the action is null or a non-expenditure. But I agree it's not an internally consistent function.

Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:246]: I could add `return actions[0];` at the end of the first branch if no expenditure sigs are found, so at least the typing will be consistent. Alternatively I could add a comment explaining the intended behavior, which would be my preference.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:1]: I guess that would work, something like `VotingReputationExtensionInterface` or something like that which exposes the extension interface functions. It's a little clunky but on balance the contract size savings are probably worth it.

File: contracts/extensions/votingReputation/VotingReputation.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputation.sol:434]: I get that, but those values aren't used in cases where these sigs appear. So they're "undefined" in that sense. I could explicitly set them to 0 if you prefer.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:178]: I agree that some branching is needed to handle multicall, but I'd ideally like to do it in a way which avoids needing to add backwards-compatibility logic.

File: contracts/extensions/votingReputation/VotingReputationStaking.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStaking.sol:76]: I'll have to prevent the locking as well.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:246]: > rather it's not clear that this function assumes that either the action is an expenditure function or that the contents contain an expenditure function.

If this is the case, can we not `require` it?

File: contracts/extensions/votingReputation/VotingReputation.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputation.sol:434]: I would vastly prefer that, yes :smile:

File: contracts/extensions/votingReputation/VotingReputation.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputation.sol:269]: I think I agree with this analysis, but I still don't think I'm comfortable with this. In general with this PR, there's a fair amount of weird behaviour, or edge cases like this one that don't _currently_ matter, but by it existing it becomes far too easy to introduce bugs down the line, as well as more likely there being a bug that we've not spotted, so I'd like to make things  as consistent as possible.

In terms of cleanly identifying 'old' motions, wee could store `motionCount` when the contract was updated to v10, and key based off the combination of that and `motionId` to determine whether a motion needs the 'old' functionality or not?

Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputation.sol:269]: Sure. I'll re-iterate that I think this extension should be sunset in the near future and replaced with a V2, but until then I'm happy to implement the backwards-compatibility in this way.

**Comment by kronosapiens** [comment]
@area updated in response to your feedback, waiting on your response on some open questions.

File: contracts/extensions/votingReputation/VotingReputationStaking.sol
Comment (area) [inline contracts/extensions/votingReputation/VotingReputationStaking.sol:79]: Okay, so we're saying that:

V9 motions, if they do not relate to expenditures, can complete as normal.

If they do relate to expenditures:
* In the voting stage: can proceed to completion (with old functionality)
* In the staking stage, if YAY not yet staked: cannot proceed further. Otherwise, can move to completion

I'm not wild that this allows people to stake on the `YAY` side, but just not to completion, nor that whether something can move out of staking is determined by which side has staked how much so far. Is a clearer demarcation possible?

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:216]: I actually don't think that would be correct behavior here. If someone creates an "invalid" multicall in version 9, and then upgrades, the motion would be stuck in perpetuity (since the downstream functions would always error).

I think our options are:

- Allow invalid multicalls created before the upgrade to be executed
- Prevent invalid multicalls from being executed by returning "finalized" instead of "finalizable"

File: contracts/extensions/votingReputation/VotingReputationStaking.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStaking.sol:79]: Sure, I can re-write that check to occur in all calls to `StakeMotion`. My motivation for doing it the way I did was to reduce the gas overhead of the check, since it would only be done in some subset of staking calls.

Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStaking.sol:79]: Ultimately the goal here is to prevent v9 motions from locking the expenditure after we upgrade.

We have to also consider the scenario where the `YAY` side has already staked, and then the extension was upgraded. In that case, the expenditure will have already been locked, and so we should allow `NAY` staking to continue (otherwise upgrading the extension is a free win to the `YAY` side).

If the `YAY` side hasn't fully staked, then we should probably disallow staking entirely on the motion, on both sides.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:217]: I've implemented this to tag any invalid summaries which would otherwise be "finalizable" as "finalized".

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:254]: It was a stylistic choice, as I didn't like how verbose most of the calls became when I had to include `getTarget`. Essentially I moved the calls to `getTarget` into the function body, whereas before it had been the responsibility of the calling function.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:301]: I added an additional test showing that this functionality would work for arbitrary targets, and realized that it needed to be implemented as a try/catch to handle all cases.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:254]: I misread the diff, I understand and agree with putting this call here.

> The changes to the test reflect the fact that an altTarget of 0x0 is interpreted as the colony address.

Right, but the colony address was already being passed in, so they could have also been left as-is, right?

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:301]: Do we _want_ it to work for arbitrary targets though? Surely the behaviour we wish to encourage is for calls going to contracts without `getCapabilityRoles` (the implication they are external, non-colony contracts) is to be an arbitrary transaction that the colony executes?

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:254]: Could have, but that would have been a hack as we don't allow `altTarget` to be the colony. So passing in `ZERO_ADDRESS` is technically the correct usage, since it gets converted into the colony address by `getTarget`. That whole functionality is a bit odd frankly (was meant as a gas optimisation), but so it goes.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol]: In which case, the comments in `getActionSummary` need updating.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:301]: That's a fair point. I believe that this behavior exists in the current version (an alternative target without `getCapabilityRoles` would result in an error), but I read that as a bug more than a feature. So we should decide what the intended behavior should be. In principle I don't see any reason to limit the alternative targets.

File: contracts/extensions/votingReputation/VotingReputationStaking.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStaking.sol]: > If the YAY side hasn't fully staked, then we should probably disallow staking entirely on the motion, on both sides.

I'd be happier with this functionality (for motions that would lock expenditures)

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:301]: I mean, I think there's a strong argument that it should only be able to target its colony and that colony's extensions.

Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:301]: I hear that, but if that's the case then I would argue we should implement this via a check to `isExtension`, as using `getCapabilityRoles` seems like a bit of a hack. Someone could create a contract implementing `getCapabilityRoles` to get around that constraint. Given that, I would say there's no reason to add that restriction until is becomes necessary.

Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:301]: I agree someone could implement `getCapabilityRoles`; both would be needed though. One is to check that it's the colony/an extension, and the other is to check what permissions the function call needs.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol]: I'm talking about the comments in `getActionSummary` that say `Invalid multicall, caller should error`.

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:301]: Following on our call on Friday, we will keep the existing functionality (function throws) for now and explicitly incorporate an `isExtension` check in a future PR.

File: contracts/testHelpers/VotingReputationV9.sol
Comment (kronosapiens) [reply contracts/testHelpers/VotingReputationV9.sol:1]: Noting this here: https://github.com/JoinColony/colonyNetwork/issues/1148

**Review by kronosapiens — COMMENTED** [review]
Just gave this another review, added some comments and did some light refactoring. Should be good for a final review and merge.

**Review by area — COMMENTED** [review]
If pre-upgrade I make a motion changing e.g. a finalized timestamp, and upgrade, I can still create motions affecting that expenditure if it's to change a slot payout. Specifically that scenario can be caught, though I don't see a way that the reverse could be (i.e. making a slot-motion first, upgrading, and then making one to affect the timestamp), nor if both motions were for different slots. I'm not sure what we want to do about it - I could see it meriting a conversation with Arren/Jack.

Post-upgrade, if a motion was about slot, the slot's delay is not correctly reset on completion.

I've added tests for these two cases that currently fail.

File: contracts/extensions/votingReputation/VotingReputationDataTypes.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationDataTypes.sol:44]: I added it to the end because otherwise I worried it would have messed up the storage layout for existing motions. When we release a proper V2 we can re-organize the storage layout. Unless you don't think there's an issue?

**Comment by kronosapiens** [comment]
To clarify @areas point about the slot locking, consider the following example:

In version 9, I create two motions affecting expenditure 8, one affecting slot 1,009 and one affecting slot 7,402. I now upgrade and create a new motion affecting expenditure 8. There is no feasible way to check that there are no active v9 motions affecting slots.

So we likely need to be ok with there being some possibility of overlap during the upgrade window, where an expenditure motion can be created while a slot-level motion affecting the same expenditure is also being resolved. On the plus side, the new expenditure motion will only lock the global claim delay, so there's no possibility of collision or interference there.

Another option would be to restrict expenditure motions after the upgrade for the maximum lifecycle of a motion, but that is probably too great of a restriction.

**Comment by arrenv** [comment]
> If pre-upgrade I make a motion changing e.g. a finalized timestamp, and upgrade, I can still create motions affecting that expenditure if it's to change a slot payout. Specifically that scenario can be caught, though I don't see a way that the reverse could be (i.e. making a slot-motion first, upgrading, and then making one to affect the timestamp), nor if both motions were for different slots. I'm not sure what we want to do about it - I could see it meriting a conversation with Arren/Jack.
> 
> Post-upgrade, if a motion was about slot, the slot's delay is not correctly reset on completion.
> 
> I've added tests for these two cases that currently fail.

Is this still needing to be discussed? Is it something we can detect and handle in the UI? I.e. if a Motion was used to change a slot, then the extension was upgraded, and then someone tries to change the timestamp?

File: contracts/extensions/votingReputation/VotingReputationDataTypes.sol
Comment (area) [reply contracts/extensions/votingReputation/VotingReputationDataTypes.sol:44]: There shouldn't be, based on my understanding. How do you think it's going to change the storage layout?

**Comment by area** [comment]
It can be detected in the UI, so we can prevent them creating more motions in the UI. We still have to be comfortable with whatever is possible with direct contract interaction though.

**Review by area — COMMENTED** [review]
> So we likely need to be ok with there being some possibility of overlap during the upgrade window, where an expenditure motion can be created while a slot-level motion affecting the same expenditure is also being resolved. On the plus side, since we've disallowed slot motions the new expenditure motion will always lock the global claim delay, so there's no possibility of collision or interference there (unlike the case where a v9 main expenditure motion conflicts with a new main expenditure motion).

> Another option would be to restrict expenditure motions after the upgrade for the maximum lifecycle of a motion, but that is probably too great of a restriction.

I agree that the last suggestion here is too much, especially with the possibility of escalations increasing the lifecycle.

The best idea I have (in conjunction with checking on the frontend) is to have a public function that could be called to prove that there's multiple pending motions affecting the same expenditure, causing the newest one to be moved to finalized. 

I've also added another test for an edge case that fails that (I believe) shouldn't.

File: contracts/extensions/votingReputation/VotingReputationDataTypes.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationDataTypes.sol:44]: If we insert it into the existing storage layout, wouldn't it push existing data down a slot? The only place I could see it going is after `altTarget`, where it would get packed after existing data. Do you agree?

File: contracts/extensions/votingReputation/VotingReputationStorage.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStorage.sol:477]: I just reintroduced the prior function as-was and kept the new version which I would want to maintain longer-term. I would rather keep them separate since `createClaimDelay` action is only for backwards-compatibility.

File: contracts/extensions/votingReputation/VotingReputationStaking.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStaking.sol:85]: Valid point. I can add a `shortcircuit` return boolean from `lockExpenditure` and skip the timestamp setting if the expenditure has been locked.

**Comment by kronosapiens** [comment]
@area your latest commits look good to me!

File: contracts/extensions/votingReputation/VotingReputationDataTypes.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationDataTypes.sol:43]: I've update the layout as you requested and tested to make sure upgrades don't mess up the layout.

File: contracts/extensions/votingReputation/VotingReputationStaking.sol
Comment (kronosapiens) [reply contracts/extensions/votingReputation/VotingReputationStaking.sol:57]: I am happy to document that v9 motions cannot be escalated after an upgrade.

**Review by area — APPROVED** [review]
I am approving, but leaving it up to you, @kronosapiens, whether you want to merge or not based on the following:

* I think the upgrade tests can be better as discussed (and also as discussed, am not suggesting this would hold up merging).
* I think `unlockExpenditure` could be split in to two functions, one for <=v9 and one for >=v10 - it's called from two different branches, based on whether it's an old expenditure or not, and then `unlockExpenditure` does the same branch again, which seems redundant.
* I [still](https://github.com/JoinColony/colonyNetwork/pull/1135#discussion_r1286876895) think that some function names are pretty bad - in addition to what I mention there, consider the functions `getExpenditureAction` and `createExpenditureAction`, and how different they are.
* I think the behaviour of a motion created in v9 that is subsequently deemed an invalid multicall is weird. It can be staked against, voted on, and escalated, but then just jumps to a `finalized` state? Why are we not allowing it to be finalized?

**Pull request merged**
